commands/service_publish.go

```diff
@@ -0,0 +1,105 @@
+package commands
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/mesg-foundation/core/ipfs"
+	"github.com/mesg-foundation/core/service/importer"
+	"github.com/spf13/cobra"
+)
+
+type servicePublishCmd struct {
+	baseCmd
+
+	path string
+	ipfs *ipfs.IPFS
+
+	e ServiceExecutor
+}
+
+type marketplaceData struct {
+	Version int `json:"version,omitempty"`
+	Service struct {
+		Deployment struct {
+			Source string `json:"source,omitempty"`
+		} `json:"deployment,omitempty"`
+	} `json:"service,omitempty"`
+	Definition *importer.ServiceDefinition `json:"definition,omitempty"`
+	Readme     string                      `json:"readme,omitempty"`
+}
+
+func newServicePublishCmd(e ServiceExecutor) *servicePublishCmd {
+	c := &servicePublishCmd{e: e}
+	c.cmd = newCommand(&cobra.Command{
+		Use:     "publish",
+		Short:   "Publish a service on the MESG Marketplace",
+		Example: `mesg-core service publish PATH_TO_SERVICE`,
+		PreRunE: c.preRunE,
+		RunE:    c.runE,
+		Args:    cobra.MaximumNArgs(1),
+		Hidden:  true, // TODO: Remove when this feature is finished
+	})
+	return c
+}
+
+func (c *servicePublishCmd) preRunE(cmd *cobra.Command, args []string) error {
+	c.path = getFirstOrDefault(args)
+	c.ipfs = ipfs.New()
+	return nil
+}
+
+func (c *servicePublishCmd) runE(cmd *cobra.Command, args []string) error {
```

Comment:
This whole function should be moved away from here. Commands pkg is only for input/output, so this should be rather in `commands/provider`

Reply:
but this command for now has no need of the api and any core client so it doesn't make sense yet to move it into the serviceprovider for me

Reply:
It's not about API or Core client.

The package commands shouldn't do any operation expect input/output (doesn't matter if it need contact via api, http or any other service like ipfs).

This is why providers are, to avoid putting logic inside commands package (or to put as less logic as possible)

Reply:
done